Https imports (#36328)
nodejs/node: doc/api/esm.md

<i id="esm_experimental_loaders"></i>

## HTTPS and HTTP imports

> Stability: 1 - Experimental

Importing network based modules using `https:` and `http:` is supported under
the `--experimental-https-modules` flag. This allows web browser-like imports to
work in Node.js with a few differences due to application stability and security
concerns that are different when running in a privileged environment instead of
a browser sandbox.

### HTTP is limited to loopback addresses

`http:` is vulnerable to man in the middle attacks and is not allowed to be used
for addresses outside of the IPv4 address 127.0.0.1 and the IPv6 address ::1 .
Support for `http:` is intended to be used for local development.

### Cookies are never sent to the destination server.

Support for custom headers per location such as `Cookie:` is in development.

> 

Reviewer: May want to say something also about authenticated requests... or lack thereof...

Author: I'm torn here since `https:` itself is a secure channel. I think until we have a mitigation implementation it isn't useful to bring up to average docs user.

Reviewer: I think it's enough just to say that the `Authorization` header will not be sent and that the import URL specifiers should not contain userinfo segments.

Author: reworded

> ### CORS is never checks on the destination server

Reviewer 2: ```suggestion
### CORS is never checked on the destination server
```

node b.mjs # works
```

## HTTPS and HTTP imports

> Stability: 1 - Experimental

Importing network based modules using `https:` and `http:` is supported under
the `--experimental-https-modules` flag. This allows web browser-like imports to
work in Node.js with a few differences due to application stability and security
concerns that are different when running in a privileged environment instead of
a browser sandbox.

### Imports are limited to HTTP/1

Automatic protocol negotiation for HTTP/2 and HTTP/3 is not yet supported.

### HTTP is limited to loopback addresses

> `http:` is vulnerable to man in the middle attacks and is not allowed to be used

Reviewer 3: ```suggestion
`http:` is vulnerable to man-in-the-middle attacks and is not allowed to be used
```

for addresses outside of the IPv4 address 127.0.0.1 and the IPv6 address ::1 .
Support for `http:` is intended to be used for local development.

### Authentication is never sent to the destination server.

`Authorization`, `Cookie`, and `Proxy-Authorization` headers are not sent to the
server. Avoid including user info in parts of imported URLs. A security model
for safely using these on the server is being worked on.

### CORS is never checked on the destination server

CORS is designed to allow a server to limit the consumers of an API to a
specific set of hosts. This is not supported as it does not make sense for a
server-based implementation.

### Cannot load non-network dependencies

These modules cannot access other modules that are not over `http:` or `https:`.

> Usage of parameters and initializers can get around this:

Reviewer 4: This feels a little overly restrictive to me. Basically it rules out a lot of use cases. Ideally if we have a more fine-grained permissions model, it would be possible to scope what "scopes" can import what protocols / hosts / paths.

Author: Policies can be used for exceptional cases and are the way to give permissions to resolution generally.

Reviewer 4: I wouldn't call these cases exceptional I would call them the norm. JSPM Node.js import maps rely on this significantly.

It seems arbitrary to make this policy decision over any other.... ideally policies can tackle these restrictions generally to begin with making this an easy "one line rule" to add.

Author: The concern is how radical transitive dependencies can be intercepted or change ownership. I don't think it is arbitrary. With opt-in you avoid having this problem be introduced accidentally. Even with import maps they are similar to policies in that they override normal resolution behavior. The goal is to allow such imports across protocols, not to do so without any differences from the web which has a different set of security concerns. Adding a single line could in theory be possible with patterns if we add them, but still likely shouldn't occur since arbitrary access to files and core are exactly what is mitigated.

Reviewer 4: I would argue this is in fact the primary feature of policy overall - in being able to delegate which URL "scopes" have which core package permissions. As such I hope we can treat it as a very natural thing for policy to do, and that the default scenario of "allow all" shouldn't require explicitly and manually adding 100s of items, while still putting security first.

Author: The default when using a policy must not be to allow all.

Reviewer 4: Would policy scopes for external URLs permit allowing core modules explicitly? Eg like:

```
"scopes": {
    "https://cdn.com/": {
      "dependencies": {
        "fs": true
      }
    }
  }
```

node b.mjs # works
```

## HTTPS and HTTP imports

> Stability: 1 - Experimental

Importing network based modules using `https:` and `http:` is supported under
the `--experimental-https-modules` flag. This allows web browser-like imports to
work in Node.js with a few differences due to application stability and security
concerns that are different when running in a privileged environment instead of
a browser sandbox.

### Imports are limited to HTTP/1

Automatic protocol negotiation for HTTP/2 and HTTP/3 is not yet supported.

### HTTP is limited to loopback addresses

`http:` is vulnerable to man-in-the-middle attacks and is not allowed to be used

> for addresses outside of the IPv4 address 127.0.0.1 and the IPv6 address ::1.

Reviewer 5: I feel like people are going to keep asking to expand this list. “What about all 192.168.x.x addresses?” “What about the IP address of the Docker host” (like how in Mac, `docker.for.mac.localhost` is a special hostname for a container to connect to its host). We probably don’t need to worry about this now in this PR, but we might want to consider something like a flag that turns off this restriction, or some other way that users could approve specific IP addresses or hostnames to not require https. Or maybe the existing `NODE_TLS_REJECT_UNAUTHORIZED` flag is extended to also disable this restriction.

> for addresses outside of the IPv4 address 127.0.0.1 and the IPv6 address ::1.

Author: We definitely do not want cross machine (including cross container) linkage to occur. Just loopback. Rest are security concerns. If someone wants to do a very large writeup on them and maintain that they can feel free to do so in other places but I would push back against it given existence of various things. I'd note that if a hostname does resolve to a loopback address it also has a leak of DNS queries though.

> for addresses outside of the IPv4 address 127.0.0.1 and the IPv6 address ::1.

Reviewer 2: There are a lot of reasons people use various "local" addresses. I think an allow list is the only viable option other than "http is not allowed at all".

> for addresses outside of the IPv4 address 127.0.0.1 and the IPv6 address ::1.

Author: @jsumners that can be done as a different feature request. I disagree with your ultimatum but am open to less opaque discussions in a different PR.

> for addresses outside of the IPv4 address 127.0.0.1 and the IPv6 address ::1.

Reviewer 2: I don't think I made an ultimatum. I only provided an opinion. I was stating that it is not unheard of to have many loopback addresses, or local addresses one considers safe. So in _my opinion_ an allow list or complete denial of `http:` are the best options.

> for addresses outside of the IPv4 address 127.0.0.1 and the IPv6 address ::1.

Author: @jsumners you stated "only viable option" which makes it an opaque choice, the clarification that it is two of the preferred options is much easier to discuss, thank you.

<i id="esm_experimental_loaders"></i>

## HTTPS and HTTP imports

> Stability: 1 - Experimental

Importing network based modules using `https:` and `http:` is supported under
the `--experimental-https-modules` flag. This allows web browser-like imports to
work in Node.js with a few differences due to application stability and security
concerns that are different when running in a privileged environment instead of
a browser sandbox.

### HTTP is limited to loopback addresses

`http:` is vulnerable to man in the middle attacks and is not allowed to be used
for addresses outside of the IPv4 address 127.0.0.1 and the IPv6 address ::1 .

> Support for `http:` is intended to be used for local development.

Reviewer: We should likely stress that only http1 is supported right now.

Author: should we add http2 here?

Reviewer: Eventually... http3 also, but we need to work on the transparent discovery and protocol selection work to be implemented (it's on my list but a long ways out)

### Authentication is never sent to the destination server.

`Authorization`, `Cookie`, and `Proxy-Authorization` headers are not sent to the
server. Avoid including user info in parts of imported URLs. A security model
for safely using these on the server is being worked on.

### CORS is never checked on the destination server

CORS is designed to allow a server to limit the consumers of an API to a
specific set of hosts. This is not supported as it does not make sense for a

> server based implementation.

Reviewer: ```suggestion
server-based implementation.
```

### Authentication is never sent to the destination server.

`Authorization`, `Cookie`, and `Proxy-Authorization` headers are not sent to the
server. Avoid including user info in parts of imported URLs. A security model
for safely using these on the server is being worked on.

### CORS is never checked on the destination server

CORS is designed to allow a server to limit the consumers of an API to a
specific set of hosts. This is not supported as it does not make sense for a
server-based implementation.

### Cannot load non-network dependencies

These modules cannot access other modules that are not over `http:` or `https:`.
To still access local modules while avoiding the security concern, pass in
references to the local dependencies:

```mjs
// file.mjs
import worker_threads from 'worker_threads';
import { configure, resize } from 'https://example.com/imagelib.mjs';
configure({ worker_threads });
```

```mjs
// https://example.com/imagelib.mjs
let worker_threads;
export function configure(opts) {
  worker_threads = opts.worker_threads;
}
export function resize(img, size) {
  // Perform resizing in worker_thread to avoid main thread blocking
}
```

> ### Network-based loading is not enabled by default.

Reviewer 6: ```
### Network-based loading is not enabled by default
```

for consistency with other headings

> For now, a flag is required to enable them. In the future, a

Reviewer 6: It'd probably be nice to explicitly name the flag here (and link to the CLI entry) rather than generically saying "a flag".

Reviewer 7: Flag name added.

I'm not opposed to there being a link, but no other named flags in this document are linked. It would be strange to have just the one here (and adding it to all the others is not relevant to this PR).

node b.mjs # works
```

## HTTPS and HTTP imports

> Stability: 1 - Experimental

Importing network based modules using `https:` and `http:` is supported under
the `--experimental-https-modules` flag. This allows web browser-like imports to
work in Node.js with a few differences due to application stability and security
concerns that are different when running in a privileged environment instead of
a browser sandbox.

### Imports are limited to HTTP/1

Automatic protocol negotiation for HTTP/2 and HTTP/3 is not yet supported.

### HTTP is limited to loopback addresses

`http:` is vulnerable to man-in-the-middle attacks and is not allowed to be used

> for addresses outside of the IPv4 address 127.0.0.1 and the IPv6 address ::1.

Reviewer 8: Per https://github.com/nodejs/node/pull/36328/commits/ae6f04742eaf8a8a19d3ab3356de33c8ffc4fa0a, the entire IPv4 loopback address block (from `127.0.0.1` to `127.255.255.255`) is now allowed. The documents here should be updated.

`Authorization`, `Cookie`, and `Proxy-Authorization` headers are not sent to the
server. Avoid including user info in parts of imported URLs. A security model
for safely using these on the server is being worked on.

### CORS is never checked on the destination server

CORS is designed to allow a server to limit the consumers of an API to a
specific set of hosts. This is not supported as it does not make sense for a
server-based implementation.

### Cannot load non-network dependencies

These modules cannot access other modules that are not over `http:` or `https:`.
To still access local modules while avoiding the security concern, pass in
references to the local dependencies:

```mjs
// file.mjs
import worker_threads from 'worker_threads';
import { configure, resize } from 'https://example.com/imagelib.mjs';
configure({ worker_threads });
```

```mjs
// https://example.com/imagelib.mjs
let worker_threads;
export function configure(opts) {
  worker_threads = opts.worker_threads;
}
export function resize(img, size) {
  // Perform resizing in worker_thread to avoid main thread blocking
}
```

### Network-based loading is not enabled by default

> For now, the `--experimental-network-imports` is required to enable them. In the
> future, a different mechanism will be used to enforce this. Opt-in is required
> to prevent transitive dependencies inadvertently using potentially mutable state
> that could affect reliability of Node.js applications.

Reviewer 5: ```suggestion
For now, the `--experimental-network-imports` flag is required to enable loading
resources over `http:` or `https:`. In the future, a different mechanism will be
used to enforce this. Opt-in is required to prevent transitive dependencies
inadvertently using potentially mutable state that could affect reliability
of Node.js applications.
```
